### PR TITLE
Backend PyTorch: Add L1 and L1+L2 regularizers

### PR DESCRIPTION
**Pull request by vl-dud**
Continuation of the PR [#1884](https://github.com/lululxvi/deepxde/pull/1884)

### CODE REVIEW

**Comment by lululxvi** [comment]
We are unifying the regularization for tensorflow and paddle, see https://github.com/lululxvi/deepxde/pull/1894 . Do you think we can also unify pytorch regularization in a more unified code?

**Comment by vl-dud** [comment]
Unfortunately, this can be a bit difficult in pytorch. The implementation options seem unnecessarily complicated to me.

File: deepxde/model.py [hunk 1]
```diff
@@ -363,13 +366,18 @@ def outputs_losses_test(inputs, targets, auxiliary_vars):
...
             decay=decay,
-            weight_decay=weight_decay,
+            weight_decay=l2_factor,
         )
 
         def train_step(inputs, targets, auxiliary_vars):
             def closure():
                 losses = outputs_losses_train(inputs, targets, auxiliary_vars)[1]
                 total_loss = torch.sum(losses)
+                if l1_factor:
```
Comment (lululxvi) [inline deepxde/model.py:376]: For l1+l2 regularization, this might not be the correct way. `weight_decay` in the optimizers is implemented not as the L2 loss function. We should only consider either L1 or L2, not L1 + L2 now.

Comment (vl-dud) [reply deepxde/model.py:373]: Fixed.

**Comment by vl-dud** [comment]
I didn't notice earlier that you implementated the NysNewtonCG optimizer. Should I add L1 regularization in `train_step_nncg`?

**Comment by lululxvi** [comment]
> I didn't notice earlier that you implemented the NysNewtonCG optimizer. Should I add L1 regularization in `train_step_nncg`?

Not this PR.

**Comment by lululxvi** [comment]
Sorry, it has been a while for this PR. Could you remind me what is the purpose of this PR?

**Comment by vl-dud** [comment]
> Sorry, it has been a while for this PR. Could you remind me what is the purpose of this PR?

This is a pytorch implementation of L1 regularization.

File: deepxde/model.py [hunk 2]
```diff
@@ -363,13 +363,18 @@ def outputs_losses_test(inputs, targets, auxiliary_vars):
...
```
Comment (lululxvi) [inline deepxde/model.py:373]: The loss should be computed in `outputs_losses`, so it will be recorded and output.

Comment (vl-dud) [reply deepxde/model.py]: Please check it now.

File: deepxde/model.py [hunk 3]
```diff
@@ -278,6 +278,15 @@ def build_loss():
...
         """pytorch"""
 
+        l1_factor, l2_factor = 0, 0
+        if self.net.regularizer is not None:
+            if self.net.regularizer[0] == "l1":
+                l1_factor = self.net.regularizer[1]
+            elif self.net.regularizer[0] == "l2":
+                l2_factor = self.net.regularizer[1]
+            else:
+                raise ValueError(f"Unknown regularizer name: {self.net.regularizer[0]}")
```
Comment (lululxvi) [inline deepxde/model.py:288]: raise NotImplementedError

Comment (vl-dud) [reply deepxde/model.py]: Fixed

File: deepxde/model.py [hunk 4]
```diff
@@ -313,6 +322,11 @@ def outputs_losses(training, inputs, targets, auxiliary_vars, losses_fn):
             losses = losses_fn(targets, outputs_, loss_fn, inputs, self, aux=aux)
             if not isinstance(losses, list):
                 losses = [losses]
+            if l1_factor:
```
[the remark this answers is not in the document]
Comment (vl-dud) [reply deepxde/model.py]: Fixed

File: deepxde/model.py [hunk 5]
```diff
@@ -313,6 +324,11 @@ def outputs_losses(training, inputs, targets, auxiliary_vars, losses_fn):
             losses = losses_fn(targets, outputs_, loss_fn, inputs, self, aux=aux)
             if not isinstance(losses, list):
                 losses = [losses]
+            if l1_factor > 0:
+                l1_loss = torch.sum(
+                    torch.stack([torch.sum(p.abs()) for p in self.net.parameters()])
+                )
+                losses.append(l1_factor * l1_loss)
```
Comment (lululxvi) [inline deepxde/model.py:331]: There is an issue with line 335 about `self.loss_weights`. `self.loss_weights` doesn't include the weight for l1 loss. If there is l1 loss, we need to add an additional weight to loss_weights.

Comment (vl-dud) [reply deepxde/model.py]: Fixed
